src/lock_free/mwcas_descriptor.cpp

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2024 Database Group, Nagoya University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// the corresponding header
+#include "dbgroup/atomic/mwcas/lock_free/mwcas_descriptor.hpp"
+
+// C++ standard libraries
+#include <atomic>
+#include <cstddef>
+
+// external libraries
+#include "dbgroup/lock/common.hpp"
+
+// local sources
+#include "dbgroup/atomic/mwcas/utility.hpp"
+
+namespace dbgroup::atomic::mwcas::lock_free
+{
+
+auto
+MwCASDescriptor::MwCAS()  //
+    -> bool
+{
+  const auto desc_addr = std::bit_cast<uint64_t>(this) | kMwCASFlag;
+  if (stat_.load(std::memory_order_seq_cst) == kUndecided) {
+    auto status = kSucceeded;
+    auto undecided = kUndecided;  // あまりよくない方法かも
+    for (size_t i = 0; i < target_cnt_; ++i) {
+      auto &target = targets_[i];
+      auto word = target.addr->compare_exchange_strong(target.old_val, desc_addr, target.fence,
+                                                       target.fence);
+      while (((word & kMwCASFlag) > 0) && word != desc_addr) {
+        auto another_desc_addr = word;
+        std::this_thread::sleep_for(kBackOffTime);
+        word = target.addr->compare_exchange_strong(target.old_val, desc_addr, target.fence,
+                                                    target.fence);
+        if (word == another_desc_addr) {
+          auto *another_desc = std::bit_cast<MwCASDescriptor *>(another_desc_addr & ~kMwCASFlag);
+          another_desc->MwCAS();
+          word = target.addr->compare_exchange_strong(target.old_val, desc_addr, target.fence,
+                                                      target.fence);
+        }
+      }
+      if (word != (desc_addr) && word != target.old_val) {
+        status = kFailed;
+        break;
+      }
```

Comment:
基本的な考え方として「CASはなるべく実行しない」ということを覚えておくとよいです．99%成功すると分かっている場合は直に実行してもOKですが，大抵の場合は他のスレッドとの競合を考えないといけないので，TATAS（test-and-test-and-set）の手続きを採用してまず現在値を確認します．現在値の確認や分岐予測など余計な処理が入りますが，キャッシュの無効化に伴うキャッシュミス（が特にメニーコア環境で与える影響）に比べれば10倍マシ，という感じです．

```suggestion
      auto *addr = target.addr;
      auto word = addr->load(kSeqCst);
      while ((word & kMwCASFlag) && word != desc_addr) {
        // another MwCAS is in progress
        const auto another_desc_addr = word;
        std::this_thread::sleep_for(kBackOffTime);
        word = addr->load(kSeqCst);
        if (word != another_desc_addr) continue;  // other threads modified this field

        // a long CPU stall has been detected, so perform another MwCAS
        auto *another_desc = std::bit_cast<MwCASDescriptor *>(another_desc_addr ^ kMwCASFlag);
        another_desc->MwCASInternal();
        word = addr->load(kSeqCst);
      }

      if (word != desc_addr
          && (word != target.old_val || addr->compare_exchange_strong(word, desc_addr, kSeqCst, kSeqCst))) {
        status = kFailed;
        break;
      }
```

Reply:
すみません，凡ミスがありました．58行目以降でCASに失敗したらMwCAS自体失敗としていますが，正しくは

1. 自身が正しく `desc_addr` を埋め込んだ場合（処理継続），
2. 他のスレッドが `desc_addr` に埋め込んでくれた場合（処理継続），
3. 他のスレッドが `desc_addr` 以外の記述子を埋め込んだ場合（バックオフから再試行），および
4. 他のスレッドが `target.old_val` 以外の値を埋め込んだ場合（MwCAS失敗）

の4パターンを考慮しないといけないですね．